src/client/Client.cc

```diff
@@ -9873,7 +10039,13 @@ int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
 
   if (in->inline_version < CEPH_INLINE_NONE) {
     if (!(have & CEPH_CAP_FILE_CACHE)) {
-      onuninline.reset(new C_SaferCond("Client::_read_uninline_data flock"));
+      if (onfinish) {
+        crf_onuninline = new CRF_onuninline();
+        onuninline.reset(crf_onuninline);
+      } else {
+        cond_onuninline = new C_SaferCond("Client::_read_uninline_data flock"); 
+        onuninline.reset(cond_onuninline);
+      }
```

Comment:
It's buggy here. Not from your patches.

We shouldn't uninline the data when reading, because it won't always be granted the `Fw` caps.

Fixed in https://github.com/ceph/ceph/pull/47090.

Reply:
I've rebased onto this patch so I don't need to deal with unlinline for read.